openedx/core/lib/block_structure/transformer.py

```diff
@@ -125,4 +126,55 @@ def transform(self, usage_info, block_structure):
                 block structure, with already collected data for the
                 transformer, that is to be transformed in place.
         """
-        pass
+        raise NotImplementedError
+
+
+class FilteringTransformerMixin(BlockStructureTransformer):
+    """
+    Transformers may optionally choose to implement this mixin if their
+    transform logic can be broken apart into a lambda for optimization of
+    combined tree traversals.
+
+    For performance reasons, developers should try to implement this mixin
+    whenever possible - with this alternative, traversal of the entire block
+    structure happens only once for all transformers that implement
+    FilteringTransformerMixin.
+    """
+
+    def transform(self, usage_info, block_structure):
+        """
+        This override of the abstract base method allows FilteringTransformers
+        to define transform_block_filters instead of transform.
+        """
```

Comment:
It seems like the reason for this method is not so much to allow the override (except in the trivial sense of satisfying pylint), but to allow the filtering transformer to be run individually, and that this method won't get executed in normal operations.  Might be worth expanding the docstring to clarify.